eth/filters/api.go

```diff
@@ -355,44 +357,71 @@ func (api *FilterAPI) histLogs(notifier notifier, rpcSub *rpc.Subscription, from
 		f := api.sys.NewRangeFilter(from, head, crit.Addresses, crit.Topics)
 		logChan, errChan := f.rangeLogsAsync(ctx)
 
-		// subscribe rmLogs
+		// Subscribe the ChainReorg logs
+		// if an ChainReorg occured,
+		// we will first recv the old chain's deleted logs in descending order,
+		// and then the new chain's added logs in desecnding order
+		// see core/blockchain.go#reorg(oldHead *types.Header, newHead *types.Block) for more details
+		// if an reorg happened between `from` and `to`, we will need to think about some senarios:
+		// 1. if a reorg occurs after the currently delivered block, then because this is happening in the future, has nothing to do with the current historical sync, we can just ignore it.
+		// 2. if a reorg occurs before the currently delivered block, then we need to stop the historical delivery, and send all replaced logs instead
 		query := ethereum.FilterQuery{FromBlock: big.NewInt(from), ToBlock: big.NewInt(head), Addresses: crit.Addresses, Topics: crit.Topics}
-		rmLogsSub, err := api.events.SubscribeLogs(query, rmLogsCh)
+		reorgLogsSub, err := api.events.SubscribeLogs(query, reorgLogsCh)
 		if err != nil {
 			return 0, err
 		}
 
-		var rmLogs []*types.Log
+		var (
+			delivered  uint64
+			reorged    bool
+			reorgBlock uint64
+			reorgLogs  []*types.Log
+		)
 	FORLOOP:
 		for {
 			select {
 			case <-notifier.Closed():
-				rmLogsSub.Unsubscribe()
+				reorgLogsSub.Unsubscribe()
 				return 0, errConnectDropped
 			case err := <-rpcSub.Err(): // client send an unsubscribe request
-				rmLogsSub.Unsubscribe()
+				reorgLogsSub.Unsubscribe()
 				return 0, err
 			case log := <-logChan:
-				notifier.Notify(rpcSub.ID, &log)
-			case logs := <-rmLogsCh:
+				// We transmit all data that meets the following conditions:
+				// 1. reorg not happened
+				// 2. reorg happened but the the log is in the remainder of the currently delivered block
+				if !reorged || log.BlockNumber < reorgBlock || log.BlockNumber <= delivered {
```

Comment:
I don't understand this condition: `log.BlockNumber <= delivered`. Why did you add this?

Reply:
Ensure that all logs of the same block are sent out, avoid sending only a subset of logs.